packages/flutter/lib/src/widgets/overlay.dart

```diff
@@ -2414,55 +2418,47 @@ final class _RenderDeferredLayoutBox extends RenderProxyBox
-      final Size boxSize =
-          theaterConstraints.biggest.isFinite
-              ? theaterConstraints.biggest
-              // Accessing the theater's size is only unsafe if it is laying out the
-              // size-determining child.
-              : theater.size;
-      super.layout(BoxConstraints.tight(boxSize));
-    }
-  }
-
-  bool _theaterDoingThisLayout = false;
-  @override
-  void layout(Constraints constraints, {bool parentUsesSize = false}) {
-    assert(_needsLayout == debugNeedsLayout);
-    // Only _RenderTheater calls this implementation.
-    assert(parent != null);
-    final bool scheduleDeferredLayout = _needsLayout || this.constraints != constraints;
-    assert(!_theaterDoingThisLayout);
-    _theaterDoingThisLayout = true;
-    super.layout(constraints, parentUsesSize: parentUsesSize);
-    assert(_theaterDoingThisLayout);
-    _theaterDoingThisLayout = false;
+  /// Whether this RenderBox's layout method is currently being called by the
+  /// theater or the layoutSurrogate's [performLayout] implementation.
+  bool _doingLayoutFromTreeWalk = false;
+  void _doLayoutFrom(RenderObject treewalkParent, {required Constraints constraints}) {
+    final bool shouldAddToDirtyList = needsLayout || this.constraints != constraints;
+    assert(!_doingLayoutFromTreeWalk);
+    _doingLayoutFromTreeWalk = true;
+    super.layout(constraints);
+    assert(_doingLayoutFromTreeWalk);
+    _doingLayoutFromTreeWalk = false;
     _needsLayout = false;
     assert(!debugNeedsLayout);
-    if (scheduleDeferredLayout) {
-      final _RenderTheater parent = this.parent! as _RenderTheater;
+    if (shouldAddToDirtyList) {
+      // Instead of laying out this subtree via treewalk, adding it to the dirty
+      // list. This ensures:
+      //
+      //  1. this node will be laid out by the PipelineOwner *after* the two
+      //     nodes it depends on (the theater and the layout surrogate) are
+      //     laid out.
+      //
+      //  2. when the deferred child's child starts to do layout, the nodes
+      //     from the layout surrogate to the theater (exclusive) have finishd
+      //     doing layout, so the deferred child's child can read their sizes
+      //     and (usually) compute the paint transform of the regular child
+      //     within the Overlay.
+      //
       // Invoking markNeedsLayout as a layout callback allows this node to be
       // merged back to the `PipelineOwner`'s dirty list in the right order, if
-      // it's not already dirty. Otherwise this may cause some dirty descendants
-      // to performLayout a second time.
-      parent.invokeLayoutCallback((BoxConstraints constraints) {
+      // it's not already dirty, such that this subtree does not get laid out
+      // twice.
+      treewalkParent.invokeLayoutCallback((BoxConstraints _) {
         markNeedsLayout();
       });
     }
   }
 
+  @override
+  void layout(Constraints constraints, {bool parentUsesSize = false}) {
```

Comment:
Is it okay that we completely ignore the parentUsesSize here? Maybe leave a comment on why? And should we also assert that it actually isn't set by the parent?